File: backend/apps/accounts/permissions.py
"""
Permission classes for Admin Panel endpoints.
"""

from rest_framework import permissions


class IsSystemAdmin(permissions.BasePermission):
    """
    Allows access only to system administrators (ADMIN role or superuser).
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_admin_user


class HasAdminPanelAccess(permissions.BasePermission):
    """
    Allows access only to users who can access the Admin Panel.
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_admin_panel_access()


class IsSystemAdminOrHasPermission(permissions.BasePermission):
    """
    Base class for permission checks that allow system admins or users 
    with a specific permission flag.
    
    Usage:
        class MyViewSet(viewsets.ModelViewSet):
            permission_classes = [IsAuthenticated, CanManageUsers]
    """
    permission_name = None
    
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        if request.user.is_superuser:
            return True
            
        if request.user.is_admin_user:
            return True
            
        if self.permission_name and hasattr(request.user, self.permission_name):
            return getattr(request.user, self.permission_name)
        
        return False


class CanManageUsers(IsSystemAdminOrHasPermission):
    """
    Allows access to users who can manage other users.
    """
    permission_name = 'can_manage_users'


class CanManageDepartments(IsSystemAdminOrHasPermission):
    """
    Allows access to users who can manage departments.
    """
    permission_name = 'can_manage_departments'


class CanViewDepartmentDashboard(permissions.BasePermission):
    """
    Allows access to users who can view department dashboards.
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        if request.user.is_superuser or request.user.is_admin_user:
            return True
            
        return request.user.can_view_department_dashboard


class CanViewGlobalDashboard(permissions.BasePermission):
    """
    Allows access to users who can view global dashboards.
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        if request.user.is_superuser or request.user.is_admin_user:
            return True
            
        return request.user.can_view_global_dashboard


class CanManageConsultsGlobally(permissions.BasePermission):
    """
    Allows access to users who can manage consults globally.
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        if request.user.is_superuser:
            return True
            
        return request.user.can_manage_consults_globally


class CanManagePermissions(permissions.BasePermission):
    """
    Allows access only to users who can modify other users' permission flags.
    This is a SuperAdmin-level permission.
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        if request.user.is_superuser:
            return True
            
        return request.user.can_manage_permissions

File: frontend/src/utils/permissions.js
/**
 * Utility functions for checking user permissions and roles.
 */

/**
 * Check if user has a specific permission
 * @param {Object} user - User object from auth context
 * @param {string} permission - Permission key (e.g., 'can_manage_users')
 * @returns {boolean}
 */
export function hasPermission(user, permission) {
    if (!user) return false;
    if (user.is_superuser || user.is_admin_user) return true;
    return user.permissions?.[permission] === true;
}

/**
 * Check if user is a superadmin
 * @param {Object} user - User object from auth context
 * @returns {boolean}
 */
export function isSuperAdmin(user) {
    if (!user) return false;
    return user.is_superuser === true || user.is_admin_user === true;
}

/**
 * Check if user has admin panel access
 * @param {Object} user - User object from auth context
 * @returns {boolean}
 */
export function hasAdminPanelAccess(user) {
    if (!user) return false;
    if (isSuperAdmin(user)) return true;
    if (user.has_admin_panel_access) return true;
    
    const permissions = user.permissions || {};
    return (
        permissions.can_manage_users ||
        permissions.can_manage_departments ||
        permissions.can_view_department_dashboard ||
        permissions.can_view_global_dashboard ||
        permissions.can_manage_consults_globally ||
        permissions.can_manage_permissions
    );
}

/**
 * Get all permissions the user has
 * @param {Object} user - User object from auth context
 * @returns {Array<string>} Array of permission keys
 */
export function getUserPermissions(user) {
    if (!user) return [];
    if (isSuperAdmin(user)) {
        return [
            'can_manage_users',
            'can_manage_departments',
            'can_view_department_dashboard',
            'can_view_global_dashboard',
            'can_manage_consults_globally',
            'can_manage_permissions',
        ];
    }
    
    const permissions = user.permissions || {};
    return Object.keys(permissions).filter(key => permissions[key] === true);
}

